### PR TITLE
Support XCTestExpectation creation APIs on XCTestCase from non-main threads

### PR DESCRIPTION
**Pull request by stmontgomery**
This matches a behavior change made in the Xcode copy of XCTest and allows tests which call `XCTestCase.expectation(description:)` — as well as other `XCTestCase` APIs that return an `XCTestExpectation` — to do so from any thread. This removes a previous requirement that tests call these APIs from the main thread, and it allows existing tests using them to adopt `async` without requiring `@MainActor`.

rdar://85344643
rdar://79163972
rdar://81024086

### CODE REVIEW

File: Sources/XCTest/Public/XCTestCase.swift [hunk 1]
```diff
@@ -99,7 +97,10 @@ open class XCTestCase: XCTest {
         XCTCurrentTestCase = self
         testRun.start()
         invokeTest()
-        failIfExpectationsNotWaitedFor(_allExpectations)
+
+        let allExpectations = XCTWaiter.subsystemQueue.sync { _allExpectations }
```
Comment (stmontgomery) [inline Sources/XCTest/Public/XCTestCase.swift:101]: This is a drive-by fix to an apparently un-synchronized read of the `_allExpectations` property

**Comment by stmontgomery** [comment]
@swift-ci Please test

**Review by briancroom — APPROVED** [review]
Nice, thank you for taking care of this!
